### PR TITLE
fix-NXP-20375-marshaller-inspector-context-NPE

### PR DESCRIPTION
**Pull request by ataillefer**

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/nuxeo/nuxeo/pull/216" target="_blank">ReviewNinja

File: nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/registry/reflect/MarshallerInspector.java [hunk 1]
```diff
@@ -349,37 +360,62 @@ private Object getThreadInstance(RenderingContext ctx) {
     }
 
     /**
-     * Create a new instance of the marshaller.
+     * Create a new instance of the marshaller and inject the services.
```
Comment (chapurlatn) [inline nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/registry/reflect/MarshallerInspector.java]: Create a new instance of the marshaller. it injects the reduired services if the marshaller is not a singleton. If  it's a singleton, it prepares the context variable to handle thread localized context. Then it injects the given ctx.

File: nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/registry/reflect/MarshallerInspector.java [hunk 2]
```diff
@@ -90,6 +91,9 @@
 
     private Object singleton;
 
+    // a a boolean to save the service instrumentation state
+    private boolean servicesInjected = false;
```
Comment (efge) [inline nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/registry/reflect/MarshallerInspector.java]: a a -> a. Use a regular `/** */` comment here. `= false` is the default and should be ommitted.

File: nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/registry/reflect/MarshallerInspector.java [hunk 3]
```diff
@@ -303,20 +310,24 @@ private RenderingContext getRealContext(RenderingContext ctx) {
...
-            for (Field contextField : contextFields) {
-                ThreadSafeRenderingContext value;
-                try {
-                    value = (ThreadSafeRenderingContext) contextField.get(singleton);
-                } catch (IllegalArgumentException | IllegalAccessException e) {
-                    log.error("unable to create a marshaller instance for clazz " + clazz.getName(), e);
-                    return null;
+        if (!servicesInjected) {
+            synchronized (this) {
+                if (!servicesInjected) {
```
Comment (efge) [inline nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/registry/reflect/MarshallerInspector.java]: `servicesInjected` must be a volatile if you want to use double-checked locking

File: nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/registry/reflect/MarshallerInspector.java [hunk 4]
```diff
@@ -303,20 +310,24 @@ private RenderingContext getRealContext(RenderingContext ctx) {
...
                 }
-                value.configureThread(ctx);
             }
         }
+        for (Field contextField : contextFields) {
+            ThreadSafeRenderingContext value;
+            try {
+                value = (ThreadSafeRenderingContext) contextField.get(singleton);
+            } catch (IllegalArgumentException | IllegalAccessException e) {
+                log.error("unable to create a marshaller instance for clazz " + clazz.getName(), e);
```
Comment (efge) [inline nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/registry/reflect/MarshallerInspector.java]: Here and elsewhere I'd rather rethrow than return null, unless there's a good (documented) reason for invalid classes to be used.

File: nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/registry/reflect/MarshallerInspector.java [hunk 5]
```diff
@@ -303,20 +310,24 @@ private RenderingContext getRealContext(RenderingContext ctx) {
...
-                try {
-                    value = (ThreadSafeRenderingContext) contextField.get(singleton);
-                } catch (IllegalArgumentException | IllegalAccessException e) {
-                    log.error("unable to create a marshaller instance for clazz " + clazz.getName(), e);
-                    return null;
+        if (!servicesInjected) {
+            synchronized (this) {
+                if (!servicesInjected) {
+                    injectServices(singleton);
+                    servicesInjected = true;
```
Comment (efge) [inline nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/registry/reflect/MarshallerInspector.java]: The `injectServices` method already does `servicesInjected = true` so this is redundant.
But there are other callers than this one and they aren't called with a synchronized(this) that I can see to protect the `servicesInjected` update. This should be made consistent.

**Comment by efge** [comment]
👍

**Comment by troger** [comment]
👍

**Pull request closed without merging**